In [1]:
import pandas as pd

In [2]:
combined_df = pd.read_pickle('lesson_1.pkl')

1. Создайте мешок слов с помощью sklearn.feature_extraction.text.CountVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно. <br>
Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df. <br>
Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000. <br>
Исключим стоп-слова с помощью stop_words='english'. <br>
Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью CountVectorizer.get_feature_names().

In [4]:
from sklearn.feature_extraction.text import CountVectorizer

In [8]:
combined_df['tweet_stemmed_sent'] = combined_df.apply(lambda row: ' '.join(row['tweet_stemmed']), axis=1)
combined_df['tweet_lemmatized_sent'] = combined_df.apply(lambda row: ' '.join(row['tweet_lemmatized']), axis=1)

In [10]:
count_vectorizer_stem = CountVectorizer(max_df=0.9, max_features=1000, stop_words='english')

# Создаем the Bag-of-Words модель
bag_of_words_stem = count_vectorizer_stem.fit_transform(combined_df['tweet_stemmed_sent'])

# Отобразим Bag-of-Words модель как DataFrame
feature_names_stem = count_vectorizer_stem.get_feature_names()
pd.DataFrame(bag_of_words_stem.toarray(), columns = feature_names_stem)

abl  absolut  accept  account  act  action  actor  actual  ad  adapt  \
0        0        0       0        0    0       0      0       0   0      0   
1        0        0       0        0    0       0      0       0   0      0   
2        0        0       0        0    0       0      0       0   0      0   
3        0        0       0        0    0       0      0       0   0      0   
4        0        0       0        0    0       0      0       0   0      0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ..    ...   
49154    0        0       0        0    0       0      0       0   0      0   
49155    0        0       0        0    0       0      0       0   0      0   
49156    0        0       0        0    0       0      0       0   0      0   
49157    0        0       0        0    0       0      0       0   0      0   
49158    0        0       0        0    0       0      0       0   0      0   

       ...  yeah  year  yesterday  yo  yoga  york  young  youtub  yr  yummi  
0      ...     0     0          0   0     0     0      0       0   0      0  
1      ...     0     0          0   0     0     0      0       0   0      0  
2      ...     0     0          0   0     0     0      0       0   0      0  
3      ...     0     0          0   0     0     0      0       0   0      0  
4      ...     0     0          0   0     0     0      0       0   0      0  
...    ...   ...   ...        ...  ..   ...   ...    ...     ...  ..    ...  
49154  ...     0     0          0   0     0     0      0       0   0      0  
49155  ...     0     0          0   0     0     0      0       0   0      0  
49156  ...     0     0          0   0     0     0      0       0   0      0  
49157  ...     0     0          0   0     0     0      0       0   0      0  
49158  ...     0     0          0   0     0     0      0       0   0      0  

[49159 rows x 1000 columns]

In [11]:
count_vectorizer_lemm = CountVectorizer(max_df=0.9, max_features=1000, stop_words='english')

# Создаем the Bag-of-Words модель
bag_of_words_lemm = count_vectorizer_lemm.fit_transform(combined_df['tweet_lemmatized_sent'])

# Отобразим Bag-of-Words модель как DataFrame
feature_names_lemm = count_vectorizer_lemm.get_feature_names()
pd.DataFrame(bag_of_words_lemm.toarray(), columns = feature_names_lemm)

abl  absolut  accept  account  act  action  actor  actual  ad  adapt  \
0        0        0       0        0    0       0      0       0   0      0   
1        0        0       0        0    0       0      0       0   0      0   
2        0        0       0        0    0       0      0       0   0      0   
3        0        0       0        0    0       0      0       0   0      0   
4        0        0       0        0    0       0      0       0   0      0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ..    ...   
49154    0        0       0        0    0       0      0       0   0      0   
49155    0        0       0        0    0       0      0       0   0      0   
49156    0        0       0        0    0       0      0       0   0      0   
49157    0        0       0        0    0       0      0       0   0      0   
49158    0        0       0        0    0       0      0       0   0      0   

       ...  yeah  year  yesterday  yo  yoga  york  young  youtub  yr  yummi  
0      ...     0     0          0   0     0     0      0       0   0      0  
1      ...     0     0          0   0     0     0      0       0   0      0  
2      ...     0     0          0   0     0     0      0       0   0      0  
3      ...     0     0          0   0     0     0      0       0   0      0  
4      ...     0     0          0   0     0     0      0       0   0      0  
...    ...   ...   ...        ...  ..   ...   ...    ...     ...  ..    ...  
49154  ...     0     0          0   0     0     0      0       0   0      0  
49155  ...     0     0          0   0     0     0      0       0   0      0  
49156  ...     0     0          0   0     0     0      0       0   0      0  
49157  ...     0     0          0   0     0     0      0       0   0      0  
49158  ...     0     0          0   0     0     0      0       0   0      0  

[49159 rows x 1000 columns]

2. Создайте мешок слов с помощью sklearn.feature_extraction.text.TfidfVectorizer.fit_transform(). Применим его к 'tweet_stemmed' и 'tweet_lemmatized' отдельно. <br>
Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df. <br>
Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000. <br>
Исключим стоп-слова с помощью stop_words='english'. <br>
Отобразим Bag-of-Words модель как DataFrame. columns необходимо извлечь с помощью TfidfVectorizer.get_feature_names().

In [12]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [13]:
tfidf_vectorizer_stem = TfidfVectorizer(max_df=0.9, max_features=1000, stop_words='english')

# Создаем the Bag-of-Words модель
bag_of_words_stem_tfidf = tfidf_vectorizer_stem.fit_transform(combined_df['tweet_stemmed_sent'])

# Отобразим Bag-of-Words модель как DataFrame
feature_names_stem_tfidf = tfidf_vectorizer_stem.get_feature_names()
pd.DataFrame(bag_of_words_stem_tfidf.toarray(), columns = feature_names_stem_tfidf)

abl  absolut  accept  account  act  action  actor  actual   ad  adapt  \
0      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
1      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
2      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
3      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
4      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ...    ...   
49154  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49155  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49156  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49157  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49158  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   

       ...  yeah  year  yesterday   yo  yoga  york  young  youtub   yr  yummi  
0      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
1      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
2      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
3      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
4      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
...    ...   ...   ...        ...  ...   ...   ...    ...     ...  ...    ...  
49154  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49155  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49156  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49157  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49158  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  

[49159 rows x 1000 columns]

In [14]:
tfidf_vectorizer_lemm = TfidfVectorizer(max_df=0.9, max_features=1000, stop_words='english')

# Создаем the Bag-of-Words модель
bag_of_words_lemm_tfidf = tfidf_vectorizer_lemm.fit_transform(combined_df['tweet_lemmatized_sent'])

# Отобразим Bag-of-Words модель как DataFrame
feature_names_lemm_tfidf = tfidf_vectorizer_lemm.get_feature_names()
pd.DataFrame(bag_of_words_lemm_tfidf.toarray(), columns = feature_names_lemm_tfidf)

abl  absolut  accept  account  act  action  actor  actual   ad  adapt  \
0      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
1      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
2      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
3      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
4      0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
...    ...      ...     ...      ...  ...     ...    ...     ...  ...    ...   
49154  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49155  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49156  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49157  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   
49158  0.0      0.0     0.0      0.0  0.0     0.0    0.0     0.0  0.0    0.0   

       ...  yeah  year  yesterday   yo  yoga  york  young  youtub   yr  yummi  
0      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
1      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
2      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
3      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
4      ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
...    ...   ...   ...        ...  ...   ...   ...    ...     ...  ...    ...  
49154  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49155  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49156  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49157  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  
49158  ...   0.0   0.0        0.0  0.0   0.0   0.0    0.0     0.0  0.0    0.0  

[49159 rows x 1000 columns]

3. Натренируем gensim.models.Word2Vec модель на наших данных. <br>
Тренировать будем на токенизированных твитах combine_df['tweet_token'] <br>
Установим следующие параметры: size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34. <br>
Используем функцию train() с параметром total_examples равным длине combine_df['tweet_token'], количество epochs установим 20.

In [17]:
from gensim.models import Word2Vec

In [19]:
modelW2V = Word2Vec(sentences=combined_df['tweet_token'], size=200, window=5, min_count=2, sg = 1, hs = 0, negative = 10, workers= 32, seed = 34)

In [25]:
modelW2V.train(combined_df['tweet_token'], total_examples=len(combined_df['tweet_token']), epochs=20)

(9198346, 11790740)

4. Давайте немного потестируем нашу модель Word2Vec и посмотрим, как она работает. Мы зададим слово positive = "dinner", и модель вытащит из корпуса наиболее похожие слова c помощью функции most_similar. То же самое попробуем со словом "trump".

In [26]:
modelW2V.most_similar(positive=['dinner'])

<ipython-input-26-4eb59620e7ca>:1: DeprecationWarning: Call to deprecated `most_similar` (Method will be removed in 4.0.0, use self.wv.most_similar() instead).
  modelW2V.most_similar(positive=['dinner'])


[('cookout', 0.5425596833229065),
 ('bihdaydinner', 0.5368253588676453),
 ('tacotuesday', 0.5307826399803162),
 ('shawarma', 0.5306323766708374),
 ('bolognese', 0.5245441198348999),
 ('whoopppp', 0.5209702253341675),
 ('burritos', 0.5146664381027222),
 ('spaghetti', 0.510309636592865),
 ('foodblogger', 0.504651665687561),
 ('hamburger', 0.5033669471740723)]

In [27]:
modelW2V.most_similar(positive=['trump'])

<ipython-input-27-bb8954f7aab5>:1: DeprecationWarning: Call to deprecated `most_similar` (Method will be removed in 4.0.0, use self.wv.most_similar() instead).
  modelW2V.most_similar(positive=['trump'])


[('donald', 0.5439615845680237),
 ('dumptrump', 0.5216750502586365),
 ('suppoer', 0.5122147798538208),
 ('unfit', 0.5071739554405212),
 ('impeachment', 0.504953145980835),
 ('donaldtrump', 0.5035406351089478),
 ('racists', 0.5035175681114197),
 ('delegaterevolt', 0.5015782117843628),
 ('fuhered', 0.4987071454524994),
 ('crony', 0.49672985076904297)]

5. Из приведенных выше примеров мы видим, что наша модель word2vec хорошо справляется с поиском наиболее похожих слов для данного слова. Но как она это делает? Она изучила векторы для каждого уникального слова наших данных и использует косинусное сходство, чтобы найти наиболее похожие векторы (слова). <br>
Давайте проверим векторное представление любого слова из нашего корпуса, например "food".

In [29]:
modelW2V.wv['food']

array([ 3.77142966e-01,  2.27660075e-01,  1.68636307e-01,  2.37807184e-01,
        6.44512892e-01, -1.10097356e-01,  2.84651713e-03,  3.52884322e-01,
       -8.54381740e-01, -2.42713884e-01,  5.35877526e-01,  2.72659451e-01,
        5.94582915e-01,  1.25963938e+00, -6.72570527e-01,  5.30221879e-01,
        4.44198400e-01, -1.46984056e-01,  1.39530137e-01, -2.12162659e-01,
       -7.66951203e-01,  2.85268158e-01,  8.05231109e-02, -9.00436863e-02,
        6.60536587e-01, -1.70029327e-01, -2.12905496e-01, -8.70911062e-01,
        3.09051782e-01, -8.07214737e-01, -2.88528085e-01, -6.01088881e-01,
       -2.94612110e-01,  3.01395684e-01,  3.31683248e-01, -3.69889230e-01,
       -6.43737912e-02, -4.42474498e-04, -5.36690414e-01, -3.78346384e-01,
        6.33231699e-02, -4.40663993e-01, -7.12328434e-01, -1.12887407e-02,
       -3.22406203e-01, -6.30971313e-01, -2.45924413e-01,  1.01875938e-01,
        3.71494979e-01,  2.27056921e-01, -6.65389970e-02,  1.53625593e-01,
        1.00851156e-01, -

6. Поскольку наши данные содержат твиты, а не только слова, нам придется придумать способ использовать векторы слов из модели word2vec для создания векторного представления всего твита. Существует простое решение этой проблемы, мы можем просто взять среднее значение всех векторов слов, присутствующих в твите. Длина результирующего вектора будет одинаковой, то есть 200. Мы повторим тот же процесс для всех твитов в наших данных и получим их векторы. Теперь у нас есть 200 функций word2vec для наших данных. <br>
Необходимо создать вектор для каждого твита, взяв среднее значение векторов слов, присутствующих в твите. В цикле сделать:  <br> vec += model_w2v[word].reshape((1, size)) <br>
и поделить финальный вектор на количество слов в твите. <br>
На выходе должен получиться wordvec_df.shape = (49159, 200).

In [32]:
import numpy as np

In [77]:
size = 200
def calc_tweet_vec(row):
    vec = np.zeros((size,))
    cnt = 0
    for word in row['tweet_token']:
        if word in modelW2V.wv:
            vec += modelW2V.wv[word]
            cnt += 1
    if cnt > 0:
        vec /= cnt
    return vec

In [85]:
wordvec_df = combined_df.apply(calc_tweet_vec, axis=1)

In [86]:
wordvec_df.shape

(49159,)

In [92]:
for i in wordvec_df:
    assert len(i) == 200
print("2nd dimension shape = 200")

2nd dimension shape = 200


## Докрутить чат-бота (обучить лучший ft, w2v, взвесить idf). Модели сдампить и сложить в облако, ноутбук можно через гитхаб. 
# Не получилось - файлик prepared_answers.txt я не нашёл, а без него как-то иначе перестроить чат-бот я пока не готов...

In [2]:
import string
from pymorphy2 import MorphAnalyzer
from stop_words import get_stop_words
import annoy
from gensim.models import Word2Vec, FastText
import pickle
import numpy as np
from tqdm import tqdm_notebook

In [3]:
def preprocess_txt(line):
    spls = "".join(i for i in line.strip() if i not in exclude).split()
    spls = [morpher.parse(i.lower())[0].normal_form for i in spls]
    spls = [i for i in spls if i not in sw and i != ""]
    return spls

In [4]:
assert True

# Preprocess for models fitting

sentences = []

morpher = MorphAnalyzer()
sw = set(get_stop_words("ru"))
exclude = set(string.punctuation)
c = 0

with open("Otvety.txt", "r", encoding='utf-8') as fin:
    for line in tqdm_notebook(fin):
        spls = preprocess_txt(line)
        sentences.append(spls)
        c += 1
        if c > 100000:
            break

<ipython-input-4-cab8524c185c>:13: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for line in tqdm_notebook(fin):


In [5]:
sentences = [i for i in sentences if len(i) > 2]
sentences[0]

['вопрос', 'тдв', 'отдыхать', 'лично', 'советовать', 'завести']

In [6]:
modelW2V = Word2Vec(sentences=sentences, size=50, window=5, min_count=1)
modelFT  = FastText(sentences=sentences, size=50, window=5, min_count=1)

In [7]:
w2v_index = annoy.AnnoyIndex(50 ,'angular')
ft_index = annoy.AnnoyIndex(50 ,'angular')

index_map = {}
counter = 0

with open("prepared_answers.txt", "r") as f:
    for line in tqdm_notebook(f):
        n_w2v = 0
        n_ft = 0
        spls = line.split("\t")
        index_map[counter] = spls[1]
        question = preprocess_txt(spls[0])
        
        vector_w2v = np.zeros(50)
        vector_ft = np.zeros(50)
        for word in question:
            if word in modelW2V:
                vector_w2v += modelW2V[word]
                n_w2v += 1
            if word in modelFT:
                vector_ft += modelFT[word]
                n_ft += 1
        if n_w2v > 0:
            vector_w2v = vector_w2v / n_w2v
        if n_ft > 0:
            vector_ft = vector_ft / n_ft
        w2v_index.add_item(counter, vector_w2v)
        ft_index.add_item(counter, vector_ft)
            
        counter += 1
        
        if counter > 100000:
            break

w2v_index.build(10)
ft_index.build(10)

FileNotFoundError: [Errno 2] No such file or directory: 'prepared_answers.txt'

In [11]:
def get_response(question, index, model, index_map):
    question = preprocess_txt(question)
    vector = np.zeros(50)
    norm = 0
    for word in question:
        if word in model:
            vector += model[word]
            norm += 1
    if norm > 0:
        vector = vector / norm
    answers = index.get_nns_by_vector(vector, 3)
    return [index_map[i] for i in answers]

In [12]:
TEXT = "когда сменится президент"

In [13]:
get_response(TEXT, w2v_index, modelW2V, index_map)

<ipython-input-11-691ac5f908f7>:6: DeprecationWarning: Call to deprecated `__contains__` (Method will be removed in 4.0.0, use self.wv.__contains__() instead).
  if word in model:
<ipython-input-11-691ac5f908f7>:7: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  vector += model[word]


[]

In [14]:
get_response(TEXT, ft_index, modelFT, index_map)

<ipython-input-11-691ac5f908f7>:6: DeprecationWarning: Call to deprecated `__contains__` (Method will be removed in 4.0.0, use self.wv.__contains__() instead).
  if word in model:
<ipython-input-11-691ac5f908f7>:7: DeprecationWarning: Call to deprecated `__getitem__` (Method will be removed in 4.0.0, use self.wv.__getitem__() instead).
  vector += model[word]


[]